File: client/source/helpers/crypto.js
import * as CryptoJS from 'crypto-js'

const KEY_SIZE = 32

let _iv = key => {
    return key
        .split('')
        .reverse()
        .slice(0, KEY_SIZE)
        .join('')
}

let _merge = (key, size = 4) => {
    let chunks = []
    for (let i = 0; i < key.length; i += size) {
        chunks.push(parseInt(`0x${key.slice(i, i + size)}`, 16))
    }
    while (chunks.length > KEY_SIZE / size) {
        for (let i = 0; i < chunks.length / 2; ++i) {
            chunks[i] = chunks[i * 2] ^ chunks[i * 2 + 1]
        }
        chunks = chunks.slice(0, chunks.length / 2)
    }
    return chunks
        .map(chunk => chunk.toString(16))
        .join('')
        .slice(0, 32)
        .padEnd(KEY_SIZE, 'f')
}

export default class {
    static encrypt(key, data) {
        key = _merge(key)
        let iv = CryptoJS.enc.Hex.parse(_iv(key))
        key = CryptoJS.enc.Hex.parse(key)
        data = CryptoJS.AES.encrypt(data, key, { iv: iv })
        return data.toString()
    }

    static decrypt(key, data) {
        key = _merge(key)
        let iv = CryptoJS.enc.Hex.parse(_iv(key))
        key = CryptoJS.enc.Hex.parse(key)
        data = CryptoJS.lib.CipherParams.create({
            ciphertext: CryptoJS.enc.Base64.parse(data),
        })
        data = CryptoJS.AES.decrypt(data, key, { iv: iv })
        return data.toString(CryptoJS.enc.Utf8)
    }
}

File: api/source/components/crypto.py
from functools import reduce
from base64 import b64encode, b64decode
from binascii import a2b_hex
from Crypto.Cipher import AES

from base import Component


class Crypto(Component):
    KEY_SIZE = 32

    def encrypt(self, key, data):
        data = self.__pad(data)
        key = self.__merge(key)
        iv = a2b_hex(self.__iv(key))
        key = a2b_hex(key)
        cipher = AES.new(key, AES.MODE_CBC, iv, segment_size=128)
        data = cipher.encrypt(data)
        data = b64encode(data)
        return data.decode('utf-8')

    def decrypt(self, key, data):
        key = self.__merge(key)
        iv = a2b_hex(self.__iv(key))
        key = a2b_hex(key)
        data = b64decode(data)
        cipher = AES.new(key, AES.MODE_CBC, iv, segment_size=128)
        data = cipher.decrypt(data)
        return self.__unpad(data)

    @staticmethod
    def __merge(key, size=4):
        chunks = [
            int(key[i:i+size], 16)
            for i in range(0, len(key), size)
        ]
        while len(chunks) > Crypto.KEY_SIZE / size:
            for i in range(0, int(len(chunks) / 2)):
                chunks[i] = chunks[i * 2] ^ chunks[i * 2 + 1]
            chunks = chunks[:int(len(chunks) / 2)]
        return reduce(
            lambda accum, chunk: accum + hex(chunk)[2:],
            chunks,
            '',
        )[:Crypto.KEY_SIZE].ljust(Crypto.KEY_SIZE, 'f')

    @staticmethod
    def __iv(key):
        return key[::-1][:Crypto.KEY_SIZE]

    @staticmethod
    def __pad(data, size=16):
        length = len(data)
        padsize = size - (length % size)
        return data.ljust(length + padsize, chr(padsize))

    @staticmethod
    def __unpad(data, size=16):
        return data[:len(data) - data[-1]]
